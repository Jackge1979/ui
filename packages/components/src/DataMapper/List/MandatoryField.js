import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

export default function MandatoryField({ element, dataKey, rowDataGetter, classNameProvider }) {
	const data = rowDataGetter.getData(element, dataKey);
	const label = data[0];
	const mandatory = data[1];
	const className = classnames(classNameProvider.get(element, dataKey));
	return (
		<div className={`comp-list-row-data mandatory-field ${className}`}>
			<div className={`mandatory-field-label`}>{label}</div>
			<div className={`mandatory-field-info`}>{mandatory ? '*' : ''}</div>
		</div>
	);
}

MandatoryField.propTypes = {
	element: PropTypes.object,
	dataKey: PropTypes.string,
	classNameProvider: PropTypes.object,
	rowDataGetter: PropTypes.object,
};
