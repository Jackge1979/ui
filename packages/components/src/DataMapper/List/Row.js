import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

function renderRowData(element, key, data, classNameProvider) {
	return (
		<div className={`comp-list-row-data ${classnames(classNameProvider.get(element, key))}`}>
			{data}
		</div>
	);
}

export default class Row extends Component {
	constructor(props) {
		super(props);
		this.handleMouseEnter = this.handleMouseEnter.bind(this);
		this.handleMouseLeave = this.handleMouseLeave.bind(this);
		this.updateElementRef = this.updateElementRef.bind(this);
	}

	componentDidMount() {
		if (this.elementRef != null) {
			this.elementRef.addEventListener('mouseenter', this.handleMouseEnter);
			this.elementRef.addEventListener('mouseleave', this.handleMouseLeave);
		}
	}

	componentWillUnmount() {
		if (this.elementRef != null) {
			this.elementRef.removeEventListener('mouseenter', this.handleMouseEnter);
			this.elementRef.removeEventListener('mouseleave', this.handleMouseLeave);
		}
	}

	handleMouseEnter() {
		this.props.onEnterElement(this.props.element);
	}

	handleMouseLeave() {
		this.props.onLeaveElement(this.props.element);
	}

	updateElementRef(ref) {
		this.elementRef = ref;
	}

	render() {
		const {
			element,
			classNameProvider,
			dataKeys,
			rowDataGetter,
			onClick,
			onDoubleClick,
		} = this.props;
		return (
			<div
				className={`comp-list-row ${classnames(classNameProvider.get(element))}`}
				onClick={onClick}
				onDoubleClick={onDoubleClick}
				ref={this.updateElementRef}
				data-id={rowDataGetter.getData(element, 'id')}
			>
				{dataKeys.map(key =>
					renderRowData(element, key, rowDataGetter.getData(element, key), classNameProvider),
				)}
			</div>
		);
	}
}

Row.propTypes = {
	element: PropTypes.object,
	classNameProvider: PropTypes.func,
	dataKeys: PropTypes.array,
	rowDataGetter: PropTypes.func,
	onClick: PropTypes.func,
	onDoubleClick: PropTypes.func,
	onEnterElement: PropTypes.func,
	onLeaveElement: PropTypes.func,
};
