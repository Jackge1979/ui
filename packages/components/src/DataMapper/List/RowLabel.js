import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

export default function RowLabel({ element, dataKey, rowDataGetter, classNameProvider }) {
	const data = rowDataGetter.getData(element, dataKey);
	return (
		<div
			className={`comp-list-row-data row-label ${classnames(
				classNameProvider.get(element, dataKey),
			)}`}
		>
			{data}
		</div>
	);
}

RowLabel.propTypes = {
	element: PropTypes.object,
	dataKey: PropTypes.string,
	classNameProvider: PropTypes.object,
	rowDataGetter: PropTypes.object,
};
