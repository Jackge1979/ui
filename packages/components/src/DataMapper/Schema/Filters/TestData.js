export const Filters = {
	WITH: true,
	WITHOUT: false,
};

export const element1 = {
	id: 'elem_1',
	name: 'Firstname',
	type: 'string',
	description: 'bla bla bla',
	mandatory: false,
};

export const element2 = {
	id: 'elem_2',
	name: 'Lastname',
	type: 'string',
	description: 'bla bla bla',
	mandatory: true,
};

export const element3 = {
	id: 'elem_3',
	name: 'Birthday',
	type: 'date',
	description: 'bla bla bla',
	mandatory: false,
};

export const element4 = {
	id: 'elem_4',
	name: 'Address',
	type: 'address',
	description: 'bla bla bla',
	mandatory: true,
};

export const schema = {
	id: '5df4g6ds4df564s',
	name: 'schema',
	elements: [element1, element2, element3, element4],
};
