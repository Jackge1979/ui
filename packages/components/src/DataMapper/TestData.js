
export const element1 = {
	id: 'elem_1',
	name: 'Firstname',
	type: 'string',
	description: 'The firstname of the customer (optional)',
	mandatory: false,
};

export const element2 = {
	id: 'elem_2',
	name: 'Lastname',
	type: 'string',
	description: 'The lastname of the customer (mandatory)',
	mandatory: true,
};

export const element3 = {
	id: 'elem_3',
	name: 'Birthday',
	type: 'date',
	description: 'Thje birthday of the customer (optional)',
	mandatory: false,
};

export const element4 = {
	id: 'elem_4',
	name: 'Address',
	type: 'address',
	description: 'The address of the customer (mandatory)',
	mandatory: true,
};

export const schema = {
	id: '5df4g6ds4df564s',
	name: 'Customer info',
	elements: [element1, element2, element3, element4],
};
