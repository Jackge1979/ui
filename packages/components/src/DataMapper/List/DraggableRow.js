import React, { Component } from 'react';
import { DragSource, DropTarget } from 'react-dnd';
import PropTypes from 'prop-types';
import flow from 'lodash/flow';
import Row from './Row.js';
import { ItemTypes } from '../Constants';

const elementSource = {
	beginDrag(props) {
		return props.dndListener.beginDrag(props.element);
	},
	endDrag(props) {
		props.dndListener.endDrag();
	},
};

const elementTarget = {
	drop(props, monitor) {
		props.dndListener.drop(monitor.getItem(), props.element);
	},
	canDrop(props, monitor) {
		const sourceItem = monitor.getItem();
		return props.dndListener.canDrop(sourceItem, props.element);
	},
};

function collectForDragSource(connect, monitor) {
	return {
		connectDragSource: connect.dragSource(),
		isDragging: monitor.isDragging(),
	};
}

function collectForDropTarget(connect, monitor) {
	return {
		connectDropTarget: connect.dropTarget(),
		dragOver: monitor.isOver(),
	};
}

/**
 * This component adds the draggable behaviour to the Row component.
 */
class DraggableRow extends Component {

	render() {
		const {
			element,
			classNameProvider,
			dataKeys,
			onClick,
			onDoubleClick,
			rowDataGetter,
			rowRenderers,
			onEnterElement,
			onLeaveElement,
			connectDragSource,
			connectDropTarget,
		} = this.props;
		return connectDragSource(
			connectDropTarget(
				<div>
					<Row
						element={element}
						onClick={onClick}
						onDoubleClick={onDoubleClick}
						classNameProvider={classNameProvider}
						dataKeys={dataKeys}
						rowDataGetter={rowDataGetter}
						rowRenderers={rowRenderers}
						onEnterElement={onEnterElement}
						onLeaveElement={onLeaveElement}
					/>
				</div>,
			),
		);
	}
}

DraggableRow.propTypes = {
	element: PropTypes.object,
	classNameProvider: PropTypes.object,
	dataKeys: PropTypes.array,
	rowDataGetter: PropTypes.object,
	rowRenderers: PropTypes.object,
	onClick: PropTypes.func,
	onDoubleClick: PropTypes.func,
	onEnterElement: PropTypes.func,
	onLeaveElement: PropTypes.func,
	connectDragSource: PropTypes.func,
	connectDropTarget: PropTypes.func,
	dragOver: PropTypes.bool,
};

export default flow(
	DragSource(ItemTypes.ELEMENT, elementSource, collectForDragSource),
	DropTarget(ItemTypes.ELEMENT, elementTarget, collectForDropTarget),
)(DraggableRow);
