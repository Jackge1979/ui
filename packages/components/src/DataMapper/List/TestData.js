import RowLabel from './RowLabel';
import MandatoryField from './MandatoryField';

export const element1 = {
	id: 'elem_1',
	name: 'element 1',
	type: 'string',
	description: 'bla bla bla',
  mandatory: false,
};

export const element2 = {
	id: 'elem_2',
	name: 'element 2',
	type: 'string',
	description: 'This element is mandatory',
  mandatory: true,
};

export const KEYS = {
  NAME: 'name',
  TYPE: 'type',
  DESC: 'desc',
};

export const dataKeys = [KEYS.NAME, KEYS.TYPE, KEYS.DESC];

export const classNameProvider = {
  get(element, key) {
    if (element && key) {
			return `classname-of-(${element.id},${key})`;
		} else if (element) {
			return `classname-of-(${element.id})`;
		}
		return `classname-of-list`;
  },
  onDragOver(status) {
  },
};

export const rowDataGetter = {
  getId(element) {
    return element.id;
  },
  getData(element, key) {
		switch (key) {
			case KEYS.NAME:
				return {
					value: element.name,
					mandatory: element.mandatory,
				};
			case KEYS.TYPE:
				return element.type;
			case KEYS.DESC:
				return element.description;
			default:
				return `No data available for ${key}`;
		}
	},
};

export const rowRenderers = {
  getComponent(key) {
		switch (key) {
			case KEYS.NAME:
				return MandatoryField;
			default:
				return RowLabel;
		}
	},
};
