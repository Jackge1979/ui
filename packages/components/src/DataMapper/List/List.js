import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Row from './Row.js';
import DraggableRow from './DraggableRow.js';

function renderRow(
  element,
  classNameProvider,
  dataKeys,
  rowDataGetter,
  rowRenderers,
  draggable,
  dndListener,
  onClick,
  onDoubleClick,
  onEnterElement,
	onLeaveElement,
) {
  if (draggable) {
    return (
      <DraggableRow
  			key={rowDataGetter.getData(element, 'id')}
  			element={element}
  			onClick={onClick}
  			onDoubleClick={onDoubleClick}
        classNameProvider={classNameProvider}
        dataKeys={dataKeys}
        rowDataGetter={rowDataGetter}
        rowRenderers={rowRenderers}
        onEnterElement={onEnterElement}
  			onLeaveElement={onLeaveElement}
        dndListener={dndListener}
  		/>
    );
  }
  return (
    <Row
			key={rowDataGetter.getData(element, 'id')}
			element={element}
			onClick={onClick}
			onDoubleClick={onDoubleClick}
      classNameProvider={classNameProvider}
      dataKeys={dataKeys}
      rowDataGetter={rowDataGetter}
      rowRenderers={rowRenderers}
      onEnterElement={onEnterElement}
			onLeaveElement={onLeaveElement}
		/>
	);
}

export default class List extends Component {

  render() {
    const {
      classNameProvider,
      elements,
      dataKeys,
      rowDataGetter,
      rowRenderers,
      onScroll,
      draggable,
      dndListener,
      onClick,
      onDoubleClick,
      updateListNodeRef,
      onEnterElement,
			onLeaveElement,
    } = this.props;
    return (
      <div
        className={`comp-list ${classnames(classNameProvider.get())}`}
        ref={updateListNodeRef}
        onScroll={onScroll}
      >
        {elements.map(elem =>
          renderRow(
            elem,
            classNameProvider,
            dataKeys,
            rowDataGetter,
            rowRenderers,
            draggable,
            dndListener,
            onClick,
            onDoubleClick,
            onEnterElement,
      			onLeaveElement,
          )
        )}
      </div>
    );
  }
}

List.propTypes = {
	elements: PropTypes.array,
	classNameProvider: PropTypes.func,
  dataKeys: PropTypes.array,
  rowDataGetter: PropTypes.func,
  rowRenderers: PropTypes.object,
  onScroll: PropTypes.func,
	draggable: PropTypes.bool,
	onClick: PropTypes.func,
	onDoubleClick: PropTypes.func,
	onEnterElement: PropTypes.func,
	onLeaveElement: PropTypes.func,
	dndListener: PropTypes.func,
	updateListNodeRef: PropTypes.func,
};
