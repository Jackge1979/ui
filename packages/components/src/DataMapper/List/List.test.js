import React from 'react';
import renderer from 'react-test-renderer';
import List from './List';
import * as TestData from './TestData';

const elements = [TestData.element1, TestData.element2];

it('simple-list', () => {
	// create React tree
	const tree = renderer
		.create(
			<List
				elements={elements}
        classNameProvider={TestData.classNameProvider}
      	dataKeys={TestData.dataKeys}
      	rowDataGetter={TestData.rowDataGetter}
      	rowRenderers={TestData.rowRenderers}
        draggable={false}
			/>,
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
