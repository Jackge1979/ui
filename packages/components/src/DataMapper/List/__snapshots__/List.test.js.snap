// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple-list 1`] = `
<div
  className="comp-list classname-of-list"
  onScroll={undefined}
>
  <div
    className="comp-list-row classname-of-(elem_1)"
    data-id="elem_1"
    onClick={undefined}
    onDoubleClick={undefined}
  >
    <div
      className="comp-list-row-data mandatory-field classname-of-(elem_1,name)"
    >
      <div
        className="mandatory-field-label"
      >
        element 1
      </div>
      <div
        className="mandatory-field-info"
      >
        
      </div>
    </div>
    <div
      className="comp-list-row-data row-label classname-of-(elem_1,type)"
    >
      string
    </div>
    <div
      className="comp-list-row-data row-label classname-of-(elem_1,desc)"
    >
      bla bla bla
    </div>
  </div>
  <div
    className="comp-list-row classname-of-(elem_2)"
    data-id="elem_2"
    onClick={undefined}
    onDoubleClick={undefined}
  >
    <div
      className="comp-list-row-data mandatory-field classname-of-(elem_2,name)"
    >
      <div
        className="mandatory-field-label"
      >
        element 2
      </div>
      <div
        className="mandatory-field-info"
      >
        *
      </div>
    </div>
    <div
      className="comp-list-row-data row-label classname-of-(elem_2,type)"
    >
      string
    </div>
    <div
      className="comp-list-row-data row-label classname-of-(elem_2,desc)"
    >
      This element is mandatory
    </div>
  </div>
</div>
`;
